### PR TITLE
Added Ability to Seach a User/Org's Repo

### LINKED ISSUE #1111: Search repo in Orgs & User

### PR DESCRIPTION
**Pull request by the-cybersapien**
Title Says All :smile: 

This Fixes #1111

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -168,6 +168,7 @@ dependencies {
     implementation 'net.yslibrary.keyboardvisibilityevent:keyboardvisibilityevent:2.1.0'
     implementation 'com.airbnb.android:lottie:2.2.5'
     implementation 'com.firebase:firebase-jobdispatcher:0.8.2'
+    implementation 'com.android.support.constraint:constraint-layout:1.0.2'
```
Comment (k0shk0sh) [inline app/build.gradle]: why is this added if you aren't using it?

File: app/build.gradle [hunk 2]
```diff
@@ -188,4 +189,5 @@ dependencies {
 }
 
 apply plugin: 'ManifestClasspath'
-apply plugin: 'com.google.gms.google-services'
\ No newline at end of file
+apply plugin: 'com.google.gms.google-services'
+apply plugin: 'kotlin-android-extensions'
```
Comment (k0shk0sh) [inline app/build.gradle]: I wouldn't like to include this since i'm having Butterknife and looking at your PR you are mixing both having onClick from butterknife but including the layout ids statistically. 

Follow any Activity that is written in kotlin

File: app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import butterknife.OnEditorAction
+import butterknife.OnTextChanged
+import com.fastaccess.R
+import com.fastaccess.helper.AnimHelper
+import com.fastaccess.helper.InputHelper
+import com.fastaccess.helper.Logger
+import com.fastaccess.ui.modules.search.repos.SearchReposFragment
+import kotlinx.android.synthetic.main.activity_search_user.*
+
+class SearchUserActivity : AppCompatActivity() {
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: why not extending BaseActivity? Every activity in the app should be extending BaseActivity where it handles themes & alot more things!.

File: app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -112,229 +109,214 @@
...
+                <action android:name="com.google.firebase.MESSAGING_EVENT" />
             </intent-filter>
         </service>
 
         <meta-data
             android:name="io.fabric.ApiKey"
-            android:value="6ed82b6e0756853d7d782a3f547f84f9ecba217e"/>
+            android:value="6ed82b6e0756853d7d782a3f547f84f9ecba217e" />
 
+        <activity android:name=".ui.modules.search.SearchUserActivity"></activity>
```
Comment (k0shk0sh) [inline app/src/main/AndroidManifest.xml]: `/>` is enough and please if you having different code format, don't format the whole file instead just format the line you added.

File: app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    private var isFork = true
+    private lateinit var searchReposFragment: SearchReposFragment
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_search_user)
+        ButterKnife.bind(this)
+
+        searchReposFragment = SearchReposFragment.newInstance()
+        supportFragmentManager.beginTransaction()
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: that's a crash on Orientation change. 

you should only `commit` when `savedInstanceState == null`

File: app/src/main/java/com/fastaccess/ui/modules/user/UserPagerActivity.java [hunk 6]
```diff
@@ -243,6 +244,7 @@ public static Intent createIntent(@NonNull Context context, @NonNull String logi
 
     @Override public boolean onCreateOptionsMenu(Menu menu) {
         getMenuInflater().inflate(R.menu.share_menu, menu);
+        getMenuInflater().inflate(R.menu.search_menu_solo, menu);
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/user/UserPagerActivity.java]: I rather to added the search to the BottomSheet that shows filtering.

File: app/build.gradle (hunk 1 above)
Comment (the-cybersapien) [reply app/build.gradle]: Studio must've included this when I added a new Activity, will remove it.

File: app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt (hunk 3 above)
Comment (the-cybersapien) [reply app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: Extending baseActivity requires making new Presenter and View Classes, which I figured weren't really needed for such a small activity. Will change accordingly.
Should I just add Null to the presenter/View or make them according to the activity at hand?

Comment (k0shk0sh) [reply app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: then simply use `SearchUserActivity<BaseMvp.View, BasePresenter<BaseMvp.View>` and in provide Presenter = `BasePresenter<BaseMvp.View>()`

**Comment by the-cybersapien** [comment]
@k0shk0sh Made the required changes, please have a look again :smile:

File: app/src/main/java/com/fastaccess/ui/modules/profile/ProfilePagerFragment.java [hunk 7]
```diff
@@ -10,6 +10,7 @@
 import com.fastaccess.data.dao.FragmentPagerAdapterModel;
 import com.fastaccess.helper.BundleConstant;
 import com.fastaccess.helper.Bundler;
+import com.fastaccess.helper.Logger;
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/profile/ProfilePagerFragment.java]: this isn't needed anymore and can be removed.

File: app/src/main/java/com/fastaccess/ui/modules/profile/org/repos/OrgReposFragment.java [hunk 8]
```diff
@@ -148,6 +150,11 @@ public static OrgReposFragment newInstance(@NonNull String username) {
         //Not supported for org profile
     }
 
+    @Override
+    public String getLogin() {
+        return username;
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/profile/org/repos/OrgReposFragment.java]: you could let go of the username variable and let the getLogin() return `getArguments().getString(BundleConstant.EXTRA);`

File: app/src/main/java/com/fastaccess/ui/modules/profile/repos/ProfileReposFragment.java [hunk 9]
```diff
@@ -154,4 +156,9 @@ public void onSortOptionSelected(String selectedSortOption) {
     public void onSortDirectionSelected(String selectedSortDirection) {
         getPresenter().onSortDirectionSelected(selectedSortDirection);
     }
+
+    @Override
+    public String getLogin() {
+        return username;
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/profile/repos/ProfileReposFragment.java]: same goes here.

File: app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    @BindView(R.id.searchEditText) lateinit var searchEditText: FontAutoCompleteEditText
+
+    private var username = ""
+    private var searchTerm = ""
+    private var isFork = true
+    private lateinit var searchReposFragment: SearchReposFragment
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        ButterKnife.bind(this)
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: this could be removed safely, the base activity do call this.

File: app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+        searchReposFragment = SearchReposFragment.newInstance()
+        supportFragmentManager.beginTransaction()
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: please wrap this with 
`if(savedInstance == null)`

File: app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+                .replace(R.id.containerFragment, searchReposFragment, "SearchReposFragment")
+                .commit()
+
+        val args = intent.extras
+        Logger.d("savedS $args")
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: do we need this anymore?

File: app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        ButterKnife.bind(this)
+
+        searchReposFragment = SearchReposFragment.newInstance()
+        supportFragmentManager.beginTransaction()
+                .replace(R.id.containerFragment, searchReposFragment, "SearchReposFragment")
+                .commit()
+
+        val args = intent.extras
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: try rotate your phone and you will get a crash. 

intent is always null when activity orientation changes. 

use variables initialization and annotated them with `@State` like in any other place and init them only when savedInstance == null

File: app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        if (args != null) {
+            username = args.getString(USERNAME)
+            if (InputHelper.isEmpty(username))
+                throw UninitializedPropertyAccessException("Username cannot be empty")
+            searchTerm = args.getString(SEARCH_TERM)
+            searchEditText.setText(searchTerm)
+            onSearchClicked()
+        }
+
+        forkCheckBox.setOnClickListener {
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: why don't we use Butterknife for this as well?

File: app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        return true
+    }
+
+    private fun makeSearch() {
+        val query = "user:$username $searchTerm fork:$isFork"
+        Logger.d(query)
+        searchReposFragment.onQueueSearch(query)
+    }
+
+    override fun layout(): Int = R.layout.activity_search_user
```
Comment (k0shk0sh) [inline app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: I had love to see the override methods to be always on top, then public methods and then private methods like in every other activity fasthub has.

**Comment by k0shk0sh** [comment]
@the-cybersapien last few small things and this can be merged :D 

thanks for your effort really!

File: app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt (hunk 13 above)
Comment (the-cybersapien) [reply app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: I actually tried this. 
The Arguments bundle was saved on my devices.
Tried this on Android O Google Pixel, and Android N Oneplus 5.
Both worked without any crash.

Comment (k0shk0sh) [reply app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: That's strange, it should have crashed 🤔. But to be sure and safe please follow what I mentioned.

File: app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt (hunk 11 above)
Comment (the-cybersapien) [reply app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: @k0shk0sh Is there a reason to do this?
If I wrap the code
```kotlin
        searchReposFragment = SearchReposFragment.newInstance()
        supportFragmentManager.beginTransaction()
                .replace(R.id.containerFragment, searchReposFragment, "SearchReposFragment")
                .commit()
```
inside `if(savedInstanceState == null)`

I get an FC on orientation change saying that the fragment isn't initialized.
Without it, the activity works well.

Comment (k0shk0sh) [reply app/src/main/java/com/fastaccess/ui/modules/search/SearchUserActivity.kt]: your code should be like this

```
        supportFragmentManager.beginTransaction()
                .replace(R.id.containerFragment, SearchReposFragment.newInstance(), "SearchReposFragment")
                .commit()
```

fun getSearchFragment() : SearchReposFragment? = return AppHelper.getFragmentByTag(supportFragmentManager, "SearchReposFragment")

**Comment by the-cybersapien** [comment]
@k0shk0sh 
Please test-run this build commit. It is running without any problems. As I mentioned in the previous review, the Activity FCs if I surround the fragment declaration and committing part inside `savedInstanceState == null`

**Comment by k0shk0sh** [comment]
I'll be merging this however i'll refactor some stuff and tag you in so you could see them.

thanks.
